MapboxNavigation/NavigationMapView.swift

```diff
@@ -154,49 +166,168 @@ open class NavigationMapView: MGLMapView {
+                } else {
+                    polyline.attributes[isCurrentLeg] = index != 0
+                }
+                return polyline
+            }
+            
+            linesPerLeg.append(lines)
         }
         
-        let congestionSegments = Array(zip(segment, combinedCongestionLevel))
+        return MGLShapeCollectionFeature(shapes: Array(linesPerLeg.joined()))
+    }
+    
+    func shape(describingCasing route: Route, legIndex: Int?) -> MGLShape? {
+        var linesPerLeg: [MGLPolylineFeature] = []
         
-        // Merge adjacent segments with the same congestion level
-        var mergedCongestionSegments = [CongestionSegment]()
-        for seg in congestionSegments {
-            let coordinates = seg.0
-            let congestionLevel = seg.1
-            if let last = mergedCongestionSegments.last, last.1 == congestionLevel {
-                mergedCongestionSegments[mergedCongestionSegments.count - 1].0 += coordinates
+        for (index, leg) in route.legs.enumerated() {
+            let legCoordinates = Array(leg.steps.flatMap {
+                $0.coordinates
+            }.joined())
+            
+            let polyline = MGLPolylineFeature(coordinates: legCoordinates, count: UInt(legCoordinates.count))
+            if let legIndex = legIndex {
+                polyline.attributes[isCurrentLeg] = index != legIndex
             } else {
-                mergedCongestionSegments.append(seg)
+                polyline.attributes[isCurrentLeg] = index != 0
             }
+            linesPerLeg.append(polyline)
         }
         
-        // Filter out any segments with low congestion
-        let nontrivialCongestionSegments = mergedCongestionSegments.filter { $0.1 != CongestionLevel.unknown && $0.1 != CongestionLevel.low }
+        return MGLShapeCollectionFeature(shapes: linesPerLeg)
+    }
+    
+    func shape(for waypoints: [Waypoint]) -> MGLShape? {
+        var features = [MGLPointFeature]()
+        let letters = String.localizedStringWithFormat(NSLocalizedString("ALPHABET", bundle: .mapboxNavigation, value: "ABCDEFGHIJKLMNOPQRSTUVWXYZ", comment: "Format string for alphabet;")).characters.map { String(describing: $0) }
         
-        let baseLine = MGLPolylineFeature(coordinates: coordinates, count: route.coordinateCount)
-        baseLine.attributes["congestion"] = "unknown"
-        let lines = nontrivialCongestionSegments.map { (congestionSegment: CongestionSegment) -> MGLPolylineFeature in
-            let polyline = MGLPolylineFeature(coordinates: congestionSegment.0, count: UInt(congestionSegment.0.count))
-            polyline.attributes["congestion"] = String(describing: congestionSegment.1)
-            return polyline
+        for (waypointIndex, waypoint) in waypoints.enumerated() {
+            let feature = MGLPointFeature()
+            feature.coordinate = waypoint.coordinate
+            feature.attributes = [ "name": letters[waypointIndex] ]
```

Comment:
If any of the waypoints has a name, I think it would be better to label the named ones with initials and leave the unnamed ones… unlabeled.